### PR TITLE
LPD-20041

### PR DESCRIPTION
**Pull request by ChrisKian**
https://liferay.atlassian.net/browse/LPD-20041

This is one of those "small change, big impact" kind of fixes, so I apologize in advance.  The root of the problem is we are invalidating the session before running through the LogoutPre/PostAction classes, which means we don't have the user's session, so we don't know who just logged out, or from where.

The fix is fairly straight-forward; we should only invalidate the session if we are on the IdP side, or, if we are on the SP side AND the IdP session has been completely destroyed.  This ensures 1) we don't invalidate early (if the SamlIdpSpSession entry still exists) and 2) we eventually and properly invalidate the session (when we return to the SP side, through the SpSessionTerminationSamlPortalFilter class.  Basically, [this code](https://github.com/liferay-appsec/liferay-portal/compare/master...ChrisKian:LPD-20041?expand=1#diff-84c3978f7c48620abe94027732563a9fadffdcbebb783cbf02cb9e4444ee1a37R83-R84) is only reached when we are revisting the SP, after a complete and successful SLO.

It's a fairly complex issue, so please let me know if you have any questions, or if you want to discuss other solutions I've tried.  Thanks!

### CODE REVIEW

**Comment by ChrisKian** [comment]
ci:test:relevant

**Comment by ChrisKian** [comment]
Running SAML#ViewSingleLogoutAndForceAuthWithMultipleSPs locally.  Will post results.

**Comment by ChrisKian** [comment]
I cannot successfully run any poshi tests locally, but looking at the failure, it looks unrelated.  Here's the exact problem area:

> Click on 'I_AGREE' that should have a value of 'I Agree' --> FAILED

File: modules/dxp/apps/saml/saml-impl/src/main/java/com/liferay/saml/internal/servlet/filter/SpSessionTerminationSamlPortalFilter.java [hunk 1]
```diff
@@ -73,8 +75,14 @@ protected void doProcessFilter(
 			_singleLogoutProfile.terminateSpSession(
 				httpServletRequest, httpServletResponse);
 
-			_singleLogoutProfile.logout(
-				httpServletRequest, httpServletResponse);
+			SamlIdpSpSession samlIdpSpSession =
+				_samlIdpSpSessionLocalService.fetchSamlIdpSpSession(
+					samlSpSession.getSamlSpSessionId());
```
Comment (stian-sigvartsen) [inline modules/dxp/apps/saml/saml-impl/src/main/java/com/liferay/saml/internal/servlet/filter/SpSessionTerminationSamlPortalFilter.java:80]: The `SamlIdpSpSessionLocalService` manages data for the SAML IDP. This filter is executing for SP requests, so we cannot call that here.

**Comment by stian-sigvartsen** [comment]
The PR continues here: liferay-appsec/liferay-portal/pull/1672

**Pull request closed without merging**